libevmasm/ConstantOptimiser.cpp

```diff
@@ -144,30 +144,47 @@ bigint CodeCopyMethod::gasNeeded() const
 		// Data gas for copy routines: Some bytes are zero, but we ignore them.
 		bytesRequired(copyRoutine()) * (m_params.isCreation ? GasCosts::txDataNonZeroGas : GasCosts::createDataGas),
 		// Data gas for data itself
-		dataGas(toBigEndian(m_value))
+		dataGas(toCompactBigEndian(m_value, 1))
 	);
 }
 
 AssemblyItems CodeCopyMethod::execute(Assembly& _assembly) const
 {
-	bytes data = toBigEndian(m_value);
+	bytes data = toCompactBigEndian(m_value, 1);
+	assertThrow(data.size() >= 1 && data.size() <= 32, OptimizerException, "Invalid number encoding.");
 	AssemblyItems actualCopyRoutine = copyRoutine();
 	actualCopyRoutine[4] = _assembly.newData(data);
+	if (data.size() != 32)
+	{
+		// Update sizes for short encodings.
+		actualCopyRoutine[3] = u256(data.size());
+		actualCopyRoutine[5] = u256(32 - data.size());
```

Comment:
That will cause dirty higher order bits.

Reply:
Can you add a test that accesses a mapping (and thus uses the scratch space) prior to using the constant?

Reply:
```
mapping (uint => uint) test;

test[0] = uint256(-1); // should the be enough to pollute the scratch space?
```

Reply:
No, because it will compute `keccak256(0, 0)`. Something like `mapping(uint => mapping(uint => uint)) test; test[7][8] = 9;` would be better, I think. Ah and that might also be optimized away, so better to make the indices parameters...

Reply:
Ah ok to trigger a sha3 opcode for calculating the storage slot of `test[7][8]`.

Reply:
Wouldn't it be simpler adding an assembly block filling that memory slot?

Reply:
Yes that's probably easier.

Reply:
An easy way to fix it would be `DUP DUP1 MSTORE` after the mload. That should cost 3x3 gas, assuming no memory expansion is needed.

It may still be worth it if the `PUSH` we're trying to save on has <32 bytes.

Reply:
This bug is still unfixed, right? Still, the test does not fail.